rs/registry/node_provider_rewards/src/v1_rewards.rs

```diff
@@ -0,0 +1,692 @@
+        "Compute total permyriad XDR",
+        Operation::SumOps(rewards_xdr_total),
+    );
+    let rewards_xdr_no_reduction_total = logger().execute(
+        "Compute total permyriad XDR no performance penalty",
+        Operation::SumOps(rewards_xdr_no_penalty_total),
+    );
+    logger().add_entry(LogEntry::RewardsXDRTotal(rewards_xdr_total));
+
+    Rewards {
+        xdr_permyriad: rewards_xdr_total.to_u64().unwrap(),
+        xdr_permyriad_no_reduction: rewards_xdr_no_reduction_total.to_u64().unwrap(),
+    }
+}
+
+fn node_providers_rewardables(
+    nodes: &[RewardableNode],
+) -> AHashMap<PrincipalId, RewardablesWithNodesMetrics> {
+    let mut node_provider_rewardables: AHashMap<PrincipalId, RewardablesWithNodesMetrics> =
+        AHashMap::default();
+
+    nodes.iter().for_each(|node| {
+        let (rewardable_nodes, assigned_metrics) = node_provider_rewardables
+            .entry(node.node_provider_id)
+            .or_default();
+
+        let nodes_count = rewardable_nodes
+            .entry((node.region.clone(), node.node_type.clone()))
+            .or_default();
+        *nodes_count += 1;
+
+        if let Some(daily_metrics) = &node.node_metrics {
+            assigned_metrics.insert(node.clone(), daily_metrics.clone());
+        }
+    });
+
+    node_provider_rewardables
+}
+
+pub fn calculate_rewards(
+    days_in_period: u64,
+    rewards_table: &NodeRewardsTable,
+    rewardable_nodes: &[RewardableNode],
+) -> RewardsPerNodeProvider {
+    let mut rewards_per_node_provider = AHashMap::default();
+    let mut rewards_log_per_node_provider = AHashMap::default();
+    let node_provider_rewardables = node_providers_rewardables(rewardable_nodes);
+
+    for (node_provider_id, (rewardable_nodes, assigned_nodes_metrics)) in node_provider_rewardables
+    {
+        let mut assigned_multipliers: AHashMap<RegionNodeTypeCategory, Vec<Decimal>> =
+            AHashMap::default();
+        let mut nodes_multiplier_stats: Vec<NodeMultiplierStats> = Vec::new();
+        let total_rewardable_nodes: u32 = rewardable_nodes.values().sum();
+
+        logger().add_entry(LogEntry::RewardsForNodeProvider(
+            node_provider_id,
+            total_rewardable_nodes,
+        ));
+
+        for (node, daily_metrics) in assigned_nodes_metrics {
+            let (multiplier, multiplier_stats) =
+                assigned_nodes_multiplier(&daily_metrics, days_in_period);
+            logger().add_entry(LogEntry::RewardMultiplierForNode(node.node_id, multiplier));
+            nodes_multiplier_stats.push((node.node_id, multiplier_stats));
+            assigned_multipliers
+                .entry((node.region.clone(), node.node_type.clone()))
+                .or_default()
+                .push(multiplier);
+        }
+
+        let rewards =
+            node_provider_rewards(&assigned_multipliers, &rewardable_nodes, rewards_table);
+        let node_provider_log = mem::take(&mut *logger());
```

Comment:
Okay, so this is correct now b/c there can't be any interleaving, but I'm wondering... why have the "logger()" at all, which uses a thread_local?  

The thread_local is useful for global state, which implies it will persist and has utility beyond a single iteration.  Why not create a new instance of RewardsLog for each iteration?

Reply:
We could do this yes...I just wanted to avoid passing around the logger to each function in the module and also avoid creating a bigger object with all these function grouped that instantiate its own logger. However if this sounds better I can change it

Reply:
Makes sense ;) Changed it